### ISSUE TITLE
Non-cached locations may show fewer results due to Firebase persistence

### ISSUE THREAD

**Issue by mcomella** [mozilla-mobile/prox#321]
Ricardo and I were getting different search results despite having the same search radius (we override the Remote Config value in code) and same commit checkout 9a60154b18c685cb53dc9d2a7a9c8ab08027814c.

Notably, I was using a fresh install on the simulator and Ricardo was using his with data from a previous build. Ricardo was receiving fewer places (e.g. ~3 places for 1 km where I received ~20). We both disabled firebase persistence and received the larger place count (~20 places).

---
Since we do a one-time Firebase query (value?), I'd guess Firebase is just retrieving from the cache and not actually contacting the server for new data. I don't know if it'd ever update that data (e.g. we queried multiple times on Ricardo's build) without doing a long-term query that waits for new children to arrive ([FIRDataEventTypeChildAdded?](http://cocoadocs.org/docsets/FirebaseDatabase/3.1.1/Constants/FIRDataEventType.html)).

Firebase persistence **should work for Kona, given that all the data is cached**, however, I largely expect it to return only cached results with new locations.

I think the simple solution is to disable persistence so we don't have to spend time investigating what works and what doesn't work with Firebase.

@jhugman is ^ feasible? Will it cause server problems? Do we perceive too much data use on device?

**Comment by jhugman** [comment]
As far as I understand the Firebase swift code, it will re-sync on location changes. I can't speak to the deltas of behaviour you're seeing (different builds, with potentially different versions of the server, and potentially different caching strategies). 

I would expect [Firebase persistence](https://firebase.google.com/docs/database/ios/offline-capabilities) to be fairly straightforward, and that we should expect everyone to be on a fresh install, and using the same database.

Will it cause server problems? No. The uni-directional dataflow means that the server will be being hit, whatever. It will cause more traffic from firebase for people repeatedly opening the app in the same location, and the volumes of data for 30 or so blobs of JSON is pretty negligible.

My big obsession that firebase could solve for us are offline use cases, though we explicitly ruled them out as a feature we wanted to build for that some time ago.

**Comment by mcomella** [comment]
*tldr:*

Trade-off:
* Keep Firebase persistence
  * Pro: we (and users) use less data
  * Con: potentially limiting the experience of those outside of Hawaii
* RM Firebase persistence
  * Pro: Use more data
  * Con: Users outside of Hawaii use the full experience.

I'm on board to RM it for now.

---
> I would expect Firebase persistence to be fairly straightforward,

My confusion is when we have cached data and we query the database: does it return the cached data? Does it wait until the new data is retrieved? Does it wait for a duration for new data before returning the cached data? Does it download the new data in the background so the next cached query is updated?

My experience with Ricardo is that the new data is never retrieved. This could make sense: it appears the primary use case of Firebase is attaching a listener and waiting for new data (where cache vs. not cached is transparent), whereas we're retrieving the values once.

> As far as I understand the Firebase swift code, it will re-sync on location changes.

The issue is not for location changes but for the same location that has new data. However, as you mention:

> we should expect everyone to be on a fresh install, and using the same database.

So this issue should not be a problem for Hawaii, but potentially for people who use the app in other locations.

> I can't speak to the deltas of behaviour you're seeing (different builds, with potentially different versions of the server, and potentially different caching strategies).

We we using the same build at the same time so I believe our configurations were identical, except for the data persisting on device from previous builds.

> Will it cause server problems? No.

Then I say we remove it.

**Comment by jhugman** [comment]
It's better than you think: 

> Con: potentially limiting the experience of those outside of Hawaii

This isn't the case; your conclusion still holds however: our current design prevents us beneficially using local persistence.

For version.next, we definitely need to revisit the design of our `FirebasePlacesDatabase` and `PlacesProvider`: we've added a polling layer on top of a syncing layer, and we really just want the syncing.

**Comment by mcomella** [comment]
> > Con: potentially limiting the experience of those outside of Hawaii
> This isn't the case;

I don't understand – why not? If they have 4 places in the cache even though the remote DB has 100, I think they'd be getting 4 places. With the retries (that I don't know too much about), I think they have a lower probability of getting stuck with a small cache, but it's still possible.
